book/src/rust/rust_style_guide.md

```diff
@@ -0,0 +1,660 @@
+
+```rust
+#[cfg(feature = "...")]
+use {
+    thing1, thing2, thing3, thing4,
+};
+```
+
+**NOTE**: All imports should occur at the top of any module, and a newline should be added between the last import and the first declared object.
+
+#### Let binding
+
+
+##### Always separately bind [RAII](https://en.wikipedia.org/wiki/Resource_acquisition_is_initialization) guards
+
+
+>RAII stands for "Resource Acquisition is Initialisation". The essence of the pattern is that resource initialisation is done in the constructor of an object and finalisation in the destructor. This pattern is extended in Rust by using a RAII object as a guard of some resource and relying on the type system to ensure that access is always mediated by the guard object.
+
+Prefer
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    let guard = m.lock();
+    do_work(guard);
+    drop(guard); // unlock the lock
+    // do other work
+}
+```
+over
+
+```rust
+fn use_mutex(m: sync::mutex::Mutex<int>) {
+    do_work(m.lock());
+    // do other work
+}
+```
+
+##### Prefer conditional expressions to deferred initialization.
+
+Prefer
+
+```rust
+let foo = match bar {
+    Baz  => 0,
+    Quux => 1
+};
+```
+
+Over
+
+```rust
+let foo;
+match bar {
+    Baz  => {
+        foo = 0;
+    }
+    Quux => {
+        foo = 1;
+    }
+}
+```
+
+unless the conditions for initialization are too complex to fit into a simple conditional expression.
+
+##### Use type annotations for clarification; prefer explicit generics when inference fails.
+
+Prefer
+
+```rust
+s.iter().map(|x| x * 2)
+        .collect::<Vec<_>>()
+```
+
+Over
+
+```rust
+let v: Vec<_> = s.iter().map(|x| x * 2)
+                        .collect();
+```
+
+When the type of a value might be unclear to the reader of the code, consider explicitly annotating it in a let.
+
+On the other hand, when the type is unclear to the compiler, prefer to specify the type by explicit generics instantiation, which is usually more clear.
+
+##### Prefer immutable bindings.
+
+Use mut bindings to signal the span during which a value is mutated:
+
+```rust
+let mut v = Vec::new();
+// push things onto v
+let v = v;
+// use v immutably henceforth
+```
+
+#### Traits
+
+##### Defining Traits
+
+When defining a trait, use the following syntax:
+
+```rust
+/// DOCS
+trait Trait<T> {
+    /// DOCS
+    type Type1: Default;
+
+    /// DOCS
+    type Type2;
+
+    /// DOCS
+    const CONST_1: usize;
+
+    /// DOCS
+    const CONST_2: usize;
+
+    /// DOCS
+    fn required_method(&self, argument: Self::Type1) -> T;
+
+    /// DOCS
+    #[inline]
+    fn optional_method(&self) -> T {
+        Self::required_method(Self::Type1::default())
+    }
+}
+```
+
+Notice the ordering of components:
+
+1. Associated Types
+
+2. Associated Constants
+
+3. Methods
```

Comment:
Again with ordering, I think flexibility is more important here

Reply:
I disagree,  types and constants should be grouped together.   I don't see the need for them to be defined ad-hoc.  And if they are grouped,  the logical order is to define types first (because consts do not have any bearing on type definitions).  Then consts, because they could be consts of the types just defined,  and then methods, because they probably use the types and the constants defined.  Even if the compiler is smart enough to unscramble ad-hoc definitions, its harder to read through by a human.

Reply:
I do think types and constants should be grouped together,  rather than spread around throughout the trait definition.

Reply:
I would leave this ordering, but maybe mark it as a recommended in our style guide (potentially can do for any other ordering rules, mark them as a preferable not a mandatory)